### PR TITLE
[ TP-41199 ] - Mask secret values in auth logs

### PR DESCRIPTION
**Pull request by austin-connor**
- fix regex, tests pass now, incorporate hash recognition in regex

### CODE REVIEW

File: spec/hedgelog/scrubber_replacement_spec.rb [hunk 1]
```diff
@@ -17,6 +17,11 @@
       it { should eq 'password:*****' }
     end
 
+    context 'the input looks like a ruby hash' do
+      let(:input) { '"password"=>"bar"' }
+      it { should eq '"password"=>"*****"'}
+    end
+
```
Comment (mkeiser82) [inline spec/hedgelog/scrubber_replacement_spec.rb]: Lets update the tests here.  Use `let(:secret)` with a broad character list omitting just the limitations we know we have now.  Also add a value to the end of each test string that we can insure is still present after the masking `publicvar=foo`. 
```suggestion
    # known limitations are the characters we treat as potentially ending this key/value pair
    # ; & , and \s
    let(:secret) { 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789~`!@#$%^*()_+={}[]|\:"\'<>.?/⚠️' }

    context 'the input looks like params' do
      let(:input) { "password=#{secret}&publicvar=foo" }
      it { should eq 'password=*****&publicvar=foo' }
    end

    context 'the input looks like a key value pair' do
      let(:input) { "password:#{secret};publicvar=foo" }
      it { should eq 'password:*****;publicvar=foo' }
    end

    context 'the input looks like a ruby hash' do
      let(:input) { '{"password"=>"'+secret+'","publicvar"=>"foo"}' }
      it { should eq '{"password"=>"*****","publicvar"=>"foo"}'}
    end
```

File: spec/calculator_spec.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# in spec/calculator_spec.rb
```
Comment (mkeiser82) [inline spec/calculator_spec.rb]: Make sure to remove this file.

File: lib/calculator.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# in lib/calculator.rb
```
Comment (mkeiser82) [inline lib/calculator.rb]: Make sure to remove this file.

File: lib/hedgelog/scrub_replacement.rb [hunk 4]
```diff
@@ -3,15 +3,18 @@ class ScrubReplacement
...
-        "#{start}#{@key}#{eql}#{@replacement}#{delim}"
+        quote1 = Regexp.last_match[1]
+        key = Regexp.last_match[2]
+        quote2 = Regexp.last_match[3]
+        separator = Regexp.last_match[4]
+        quote3 = Regexp.last_match[5]
+        quote4 = Regexp.last_match[7]
+        "#{quote1}#{key}#{quote2}#{separator}#{quote3}#{@replacement}#{quote4}"
       end
     end
```
Comment (mkeiser82) [inline lib/hedgelog/scrub_replacement.rb]: With the test updates below, I am open to alternatives, but this should get things passing again:
```suggestion
    def initialize(key, replacement)
      @key = key
      @replacement = replacement
      @match_regex = /("?)#{@key}\1(=>|=|:)(\s*)("?)(.+?)\4(&|,|;|\s|$)/
    end

    def scrub_string(string)
      string.gsub!(@match_regex) do
        quote1 = Regexp.last_match[1]
        sep = Regexp.last_match[2]
        whitespace = Regexp.last_match[3]
        quote2 = Regexp.last_match[4]
        # secret = Regexp.last_match[5]
        rest = Regexp.last_match[6]
        "#{quote1}#{@key}#{quote1}#{sep}#{whitespace}#{quote2}#{@replacement}#{quote2}#{rest}"
      end
    end
```

File: spec/hedgelog/scrubber_replacement_spec.rb [hunk 5]
```diff
@@ -17,6 +17,11 @@
       it { should eq 'password:*****' }
     end
 
+    context 'the input looks like a ruby hash' do
```
[the remark this answers is not in the document]
Comment (austin-connor) [reply spec/hedgelog/scrubber_replacement_spec.rb]: Yeah I'd like to keep the single quotes as there are numerous quotation marks throughout these tests.

**Comment by mkeiser82** [comment]
LGTM

**Pull request merged**